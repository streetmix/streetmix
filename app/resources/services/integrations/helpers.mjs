import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
