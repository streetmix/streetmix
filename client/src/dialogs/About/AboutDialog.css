.about-dialog {
  max-width: 820px;

  @media only screen and (width > 768px) {
    & {
      max-height: 75vh !important;
    }
  }

  /* Used by team list and external links */
  ul {
    list-style: none;
    padding: 0;
    margin: 0;
  }

  header {
    background-color: white;
    padding: 0;
    text-align: start;
  }

  header h1 {
    display: none;
  }

  .dialog-content {
    overflow: hidden;
    display: flex;
  }
}

.about-dialog-logo {
  height: 36px;
  margin: 2.5rem 2rem 2rem;
}

.about-dialog-content {
  display: flex;
  flex-grow: 1;
}

.about-dialog-left {
  border-inline-end: 1px solid var(--color-turquoise-500);
  padding-block: 0;
  padding-inline: 2em 3.5em;
  width: min(30%, 280px);

  p:first-child {
    margin-top: 0;
    text-wrap: balance;
  }

  p {
    margin-top: 1.5em;
  }
}

.about-dialog-right {
  flex: 1;
  overflow: hidden scroll;
  padding: 0 2em 3em;

  > :first-child {
    margin-top: 0;
  }

  > h2 {
    margin-top: 1.5em;
  }
}

@media only screen and (width < 640px) {
  .about-dialog-content {
    display: block;
    overflow-y: scroll;
  }

  .about-dialog-left {
    width: 100%;
    border: 0;
  }

  .credits-team {
    max-width: 500px;
  }
}

.about-dialog-sponsors {
  margin-top: -1em !important;

  img {
    height: 40px;
  }
}

.about-dialog-sponsors li {
  display: inline-block;
  margin-block-start: 1em;
  margin-inline-end: 1.5em;
}
