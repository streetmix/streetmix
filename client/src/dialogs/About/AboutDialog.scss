.about-dialog {
  max-width: 820px;
  min-height: 660px;
  max-height: 75vh;

  // Used by team list and external links
  ul {
    list-style: none;
    padding: 0;
    margin: 0;
  }

  header {
    background-color: white;
    padding: 0;
    text-align: left;

    [dir="rtl"] & {
      text-align: right;
    }
  }

  header h1 {
    display: none;
  }

  .dialog-content {
    overflow: hidden;
    display: flex;
  }
}

.about-dialog-logo {
  height: 36px;
  margin: 2.5rem 2rem 2rem;
}

.about-dialog-content {
  display: flex;
  flex-grow: 1;
}

.about-dialog-left {
  border-right: 1px solid var(--color-turquoise-500);
  padding: 0 3.5em 0 2em;
  width: min(30%, 280px);

  [dir="rtl"] & {
    padding: 0 2em 0 3.5em;
    border-left: 1px solid var(--color-turquoise-500);
    border-right: 0;
  }

  p:first-child {
    margin-top: 0;
  }

  p {
    margin-top: 1.5em;
  }
}

.about-dialog-right {
  flex: 1;
  overflow: hidden scroll;
  padding: 0 2em 3em;

  > :first-child {
    margin-top: 0;
  }

  > h2 {
    margin-top: 1.5em;
  }
}

@media only screen and (width <= 768px) {
  .about-dialog-content {
    display: block;
    overflow-y: scroll;
  }

  .about-dialog-left {
    width: 100%;
    border: 0;
  }

  .credits-team {
    max-width: 500px;
  }
}

.about-dialog-sponsors {
  margin-top: -1em !important;

  img {
    height: 40px;
  }
}

.about-dialog-sponsors li {
  display: inline-block;
  margin-right: 1.5em;
  margin-top: 1em;

  [dir="rtl"] & {
    margin-left: 1.5em;
    margin-right: 0;
  }
}
