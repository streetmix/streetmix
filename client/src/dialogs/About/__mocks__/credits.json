{
  "team": [
    {
      "name": "Michael Scott",
      "title": "regional manager",
      "mugshotFile": "michael.jpg",
      "url": "https://dunder-mifflin.com",
      "active": true,
      "original": true
    },
    {
      "name": "Dwight Schrute",
      "title": "assistant to the regional manager",
      "mugshotFile": "dwight.jpg",
      "active": false,
      "original": true
    },
    {
      "name": "Pam Beesly",
      "title": "office administrator",
      "mugshotFile": "pam.jpg",
      "active": true
    },
    {
      "name": "Jim Halpert",
      "title": "sales",
      "mugshotFile": "jim.jpg",
      "url": "http://sabre.com",
      "active": false
    }
  ],
  "contributors": {
    "advisors": ["Jan Levinson", "Robert California"],
    "additional-illustrations": [
      "Stanley Hudson",
      "Kevin Malone",
      "Meredith Palmer"
    ],
    "additional-code": ["Angela Martin", "Oscar Martinez", "Phyllis Lapin"],
    "additional-contributors": [
      ["Kelly Kapoor", "customer service"],
      ["Toby Flenderson", "human resources"],
      "Creed Bratton"
    ],
    "special-thanks": [
      "Darryl Philbin",
      "Kelly Erin Hannon",
      ["Ryan Howard", "WUPHF"],
      ["Andy Bernard", "Cornell"],
      {
        "label": "Dunder Mifflin",
        "people": ["David Wallace", "Nellie Bertram", "Holly Flax"]
      }
    ]
  },
  "translators": {
    "ar": ["Karen Filippelli", "Charles Miner"],
    "de": ["Todd Packer", ["Gabe Lewis", "Sabre"], ["Jo Bennett", "Sabre"]]
  }
}
