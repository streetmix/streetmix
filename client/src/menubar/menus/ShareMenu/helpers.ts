import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
