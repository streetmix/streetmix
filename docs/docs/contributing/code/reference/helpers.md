---
sidebar_position: 2
---

# Helpers and utilities

:::caution Under construction

This page is a work in progress.

:::

We use the following utilities to solve common programming problems.

## Pseudo-random number generator

If you need a one-time use random number (which will never be stored), there's nothing wrong with JavaScript's built-in `Math.random()`. However, there are times where we need a **seed** for a pseudo-random number generator (PRNG). Seeds generate a random-_looking_ sequence of numbers, but they have the benefit of remaining consistent across renders. For example, this is useful on our sidewalk segment, where we want to generate a random set of pedestrians but we want those pedestrians to remain the same between page refreshes or React component updates.

We can generate seeds by calling `generateRandSeed()`, then by feeding those seeds into the `seedrandom` library:

```js title="random-example.js"
import seedrandom from "seedrandom";
import { generateRandSeed } from "./util/random";

function doThingsWithThePRNG() {
  const seed = generateRandSeed();
  const prng1 = seedrandom(seed);

  // Generate pseudo-random numbers. Each time rng() is called, it returns
  // the next pseudo-random number in the sequence.
  console.log(prng1()); // -> first pseudo-random number
  console.log(prng1()); // -> second pseudo-random number
  // .. etc

  // Create another generator with the same seed
  const prng2 = seedrandom(seed);
  console.log(prng2()); // -> same value as the first call to prng1()
}
```

When `seedrandom()` is called, it returns a function that will return a consistent sequence of numbers each time it is called, for the given `seed`. The sequence returned is local to that function.

:::note

Although the `seed` can be any type of value, like a string, `generateRandSeed()` returns only integer values so that when we store seeds in a data model we can check for type consistency.

:::

In the above example, we call `seedrandom()` twice, with the same `seed`, so that we can create the same sequence of pseudo-random numbers in two different scenarios. This example is contrived because the pseudo-random number generators are used in the same function, and in reality the `seedrandom()` call may happen in separate parts of the application.

**References**

- [Random Seeds, Coded Hints, and Quintillions](http://davidbau.com/archives/2010/01/30/random_seeds_coded_hints_and_quintillions.html) [David Bau]
- [seedrandom.js](https://github.com/davidbau/seedrandom) [GitHub]

## axios

The [Axios](https://github.com/axios/axios) library was introduced as a recommendation over native `fetch` as a way to manage requests to the Streetmix back-end API layer. Additionally, Axios Mock Adapter offers the possibility to mock all API requests for integration testing.

If you work on API-related features, consider refactoring the `fetch` calls to use the API in `.assets/scripts/util/api.js`.
