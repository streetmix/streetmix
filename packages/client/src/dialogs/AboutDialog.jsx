/**
 * About Streetmix (dialog box)
 *
 * Renders the "About" dialog box.
 *
 */
import React from 'react'
import { FormattedMessage } from 'react-intl'
import { useSelector } from 'react-redux'
import ExternalLink from '../ui/ExternalLink'
import logo from '../../images/logo_horizontal.svg'
import numoLogo from '../../images/sponsors/numo.svg'
import cfalogo from '../../images/sponsors/codeforamerica.png'
import mozlogo from '../../images/sponsors/mozilla.svg'
import Credits from './About/Credits.jsx' // Without extension, test will erroneously import .json instead
import SocialLinks from './About/SocialLinks'
import Dialog from './Dialog'
import './AboutDialog.scss'

function AboutDialog (props) {
  const offline = useSelector((state) => state.system.offline)

  return (
    <Dialog>
      {(closeDialog) => (
        <div className="about-dialog">
          <header>
            <img
              src={logo}
              alt="Streetmix (logo)"
              className="about-dialog-logo"
            />
            <h1>
              <FormattedMessage
                id="dialogs.about.heading"
                defaultMessage="About Streetmix."
              />
            </h1>
          </header>
          <div className="dialog-content">
            <div className="about-dialog-content">
              <div className="about-dialog-left">
                <p>
                  <FormattedMessage
                    id="dialogs.about.description"
                    defaultMessage="Design, remix, and share your street. Add bike paths, widen sidewalks or traffic lanes, learn how all of this can impact your community."
                  />
                </p>
                <SocialLinks />
                <h3>
                  <FormattedMessage
                    id="dialogs.about.partners"
                    defaultMessage="Partners"
                  />
                </h3>
                <ul className="about-dialog-sponsors">
                  <li>
                    <ExternalLink href="https://numo.global/">
                      <img src={numoLogo} alt="New Urban Mobility Alliance" />
                    </ExternalLink>
                  </li>
                  <li>
                    <ExternalLink href="https://codeforamerica.org/">
                      <img src={cfalogo} alt="Code for America" />
                    </ExternalLink>
                  </li>
                </ul>
                <h3>
                  <FormattedMessage
                    id="dialogs.about.sponsors"
                    defaultMessage="Sponsors"
                  />
                </h3>
                <ul className="about-dialog-sponsors">
                  <li>
                    <ExternalLink href="https://www.mozilla.org/en-US/moss/">
                      <img src={mozlogo} alt="Mozilla Open Source Support" />
                    </ExternalLink>
                  </li>
                </ul>
                {!offline && (
                  <p>
                    <a
                      href="https://streetmix.net/terms-of-service/"
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      <FormattedMessage
                        id="dialogs.about.tos-link"
                        defaultMessage="Terms of service"
                      />
                    </a>
                    <br />
                    <a
                      href="https://streetmix.net/privacy-policy/"
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      <FormattedMessage
                        id="dialogs.about.privacy-link"
                        defaultMessage="Privacy policy"
                      />
                    </a>
                  </p>
                )}
              </div>
              <div className="about-dialog-right">
                <Credits />
                <div className="credits-container">
                  <div>
                    <h3>
                      <FormattedMessage
                        id="dialogs.about.acknowledgements"
                        defaultMessage="Acknowledgements"
                      />
                    </h3>
                    <ul>
                      <li>
                        <FormattedMessage
                          id="dialogs.about.font-designed-by"
                          defaultMessage="{fontName} font designed by {fontAuthor}."
                          values={{
                            fontName: (
                              <ExternalLink href="https://manropefont.com/">
                                Manrope
                              </ExternalLink>
                            ),
                            fontAuthor: (
                              <ExternalLink href="https://gent.media/">
                                Mikhail Shiranda
                              </ExternalLink>
                            )
                          }}
                        />{' '}
                        <FormattedMessage
                          id="dialogs.about.license-label"
                          defaultMessage="(<a>License</a>)"
                          values={{
                            // eslint-disable-next-line react/display-name
                            a: (chunks) => (
                              <ExternalLink
                                href="https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=OFL"
                                title="SIL Open Font License, Version 1.1"
                              >
                                {chunks}
                              </ExternalLink>
                            )
                          }}
                        />
                      </li>
                      <li>
                        <FormattedMessage
                          id="dialogs.about.font-designed-by"
                          defaultMessage="{fontName} font designed by {fontAuthor}."
                          values={{
                            fontName: (
                              <ExternalLink href="https://hubertfischer.com/work/type-rubik">
                                Rubik
                              </ExternalLink>
                            ),
                            fontAuthor: (
                              <ExternalLink href="https://hubertfischer.com/">
                                Hubert & Fischer
                              </ExternalLink>
                            )
                          }}
                        />{' '}
                        <FormattedMessage
                          id="dialogs.about.license-label"
                          defaultMessage="(<a>License</a>)"
                          values={{
                            // eslint-disable-next-line react/display-name
                            a: (chunks) => (
                              <ExternalLink
                                href="https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=OFL"
                                title="SIL Open Font License, Version 1.1"
                              >
                                {chunks}
                              </ExternalLink>
                            )
                          }}
                        />
                      </li>
                      <li>
                        <FormattedMessage
                          id="dialogs.about.font-designed-by"
                          defaultMessage="{fontName} font designed by {fontAuthor}."
                          values={{
                            fontName: (
                              <ExternalLink href="https://delvefonts.com/fonts/overpass/">
                                Overpass
                              </ExternalLink>
                            ),
                            fontAuthor: (
                              <ExternalLink href="https://delvefonts.com/">
                                Delve Fonts
                              </ExternalLink>
                            )
                          }}
                        />{' '}
                        <FormattedMessage
                          id="dialogs.about.license-label"
                          defaultMessage="(<a>License</a>)"
                          values={{
                            // eslint-disable-next-line react/display-name
                            a: (chunks) => (
                              <ExternalLink
                                href="https://github.com/RedHatOfficial/Overpass/blob/master/OFL.txt"
                                title="SIL Open Font License, Version 1.1"
                              >
                                {chunks}
                              </ExternalLink>
                            )
                          }}
                        />
                      </li>
                      <li>
                        <FormattedMessage
                          id="dialogs.about.emoji-by"
                          defaultMessage="Emoji by {author}."
                          values={{
                            author: (
                              <ExternalLink href="https://openmoji.org/">
                                OpenMoji
                              </ExternalLink>
                            )
                          }}
                        />{' '}
                        <FormattedMessage
                          id="dialogs.about.license-label"
                          defaultMessage="(<a>License</a>)"
                          values={{
                            // eslint-disable-next-line react/display-name
                            a: (chunks) => (
                              <ExternalLink
                                href="https://creativecommons.org/licenses/by-sa/4.0/#"
                                title="Creative Commons Share Alike License 4.0 (CC BY-SA 4.0)"
                              >
                                {chunks}
                              </ExternalLink>
                            )
                          }}
                        />
                      </li>
                      <li>
                        <FormattedMessage
                          id="dialogs.about.icons-by"
                          defaultMessage="Icons by {author}."
                          values={{
                            author: (
                              <ExternalLink href="https://fontawesome.com/">
                                Font Awesome
                              </ExternalLink>
                            )
                          }}
                        />{' '}
                        <FormattedMessage
                          id="dialogs.about.license-label"
                          defaultMessage="(<a>License</a>)"
                          values={{
                            // eslint-disable-next-line react/display-name
                            a: (chunks) => (
                              <ExternalLink
                                href="https://fontawesome.com/license/free"
                                title="Creative Commons Share Alike License 4.0 (CC BY-SA 4.0)"
                              >
                                {chunks}
                              </ExternalLink>
                            )
                          }}
                        />
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button className="dialog-primary-action" onClick={closeDialog}>
            <FormattedMessage id="btn.close" defaultMessage="Close" />
          </button>
        </div>
      )}
    </Dialog>
  )
}

export default React.memo(AboutDialog)
