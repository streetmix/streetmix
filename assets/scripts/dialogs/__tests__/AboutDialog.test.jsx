/* eslint-env jest */
import React from 'react'
import { render } from '../../../../test/helpers/render'
import AboutDialog from '../AboutDialog'

jest.mock('../About/credits.json', () =>
  require('../About/__mocks__/credits.json')
)

describe('AboutDialog', () => {
  it('renders snapshot', () => {
    const { asFragment } = render(<AboutDialog />)
    expect(asFragment()).toMatchSnapshot()
  })
})
