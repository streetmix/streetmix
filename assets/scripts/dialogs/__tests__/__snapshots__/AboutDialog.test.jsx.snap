// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AboutDialog > renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="dialog-box-container"
  >
    <div
      class="dialog-box-backdrop"
    />
    <div
      class="dialog-box"
      role="dialog"
    >
      <button
        class="close"
        title="Dismiss"
      >
        <svg
          class="octicon"
        />
      </button>
      <div
        class="about-dialog"
      >
        <header>
          <img
            alt="Streetmix (logo)"
            class="about-dialog-logo"
            src="test-file-stub"
          />
          <h1>
            About Streetmix.
          </h1>
        </header>
        <div
          class="dialog-content"
        >
          <div
            class="about-dialog-content"
          >
            <div
              class="about-dialog-left"
            >
              <p>
                Design, remix, and share your street. Add bike paths, widen sidewalks or traffic lanes, learn how all of this can impact your community.
              </p>
              <ul
                class="social-links"
              >
                <li>
                  <a
                    href="https://github.com/streetmix/"
                    rel="noopener noreferrer"
                    target="_blank"
                    title="GitHub"
                  >
                    <svg
                      class="svg-inline--fa social-github fa-github"
                    />
                  </a>
                </li>
                <li>
                  <a
                    href="https://strt.mx/discord"
                    rel="noopener noreferrer"
                    target="_blank"
                    title="Discord"
                  >
                    <svg
                      class="svg-inline--fa social-discord fa-discord"
                    />
                  </a>
                </li>
                <li>
                  <a
                    href="https://urbanists.social/@streetmix"
                    rel="me noopener noreferrer"
                    target="_blank"
                    title="Mastodon"
                  >
                    <svg
                      class="svg-inline--fa social-mastodon fa-mastodon"
                    />
                  </a>
                </li>
                <li>
                  <a
                    href="https://twitter.com/streetmix"
                    rel="noopener noreferrer"
                    target="_blank"
                    title="Twitter"
                  >
                    <svg
                      class="svg-inline--fa social-twitter fa-twitter"
                    />
                  </a>
                </li>
                <li>
                  <a
                    href="https://www.instagram.com/streetmixapp/"
                    rel="noopener noreferrer"
                    target="_blank"
                    title="Instagram"
                  >
                    <svg
                      class="svg-inline--fa social-instagram fa-instagram"
                    />
                  </a>
                </li>
              </ul>
              <h3>
                Partners
              </h3>
              <ul
                class="about-dialog-sponsors"
              >
                <li>
                  <a
                    href="https://numo.global/"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <img
                      alt="New Urban Mobility Alliance"
                      src="test-file-stub"
                    />
                  </a>
                </li>
                <li>
                  <a
                    href="https://codeforamerica.org/"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <img
                      alt="Code for America"
                      src="test-file-stub"
                    />
                  </a>
                </li>
              </ul>
              <h3>
                Sponsors
              </h3>
              <ul
                class="about-dialog-sponsors"
              >
                <li>
                  <a
                    href="https://www.mozilla.org/en-US/moss/"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <img
                      alt="Mozilla Open Source Support"
                      src="test-file-stub"
                    />
                  </a>
                </li>
              </ul>
              <p>
                <a
                  href="https://streetmix.net/terms-of-service/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Terms of service
                </a>
                <br />
                <a
                  href="https://streetmix.net/privacy-policy/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Privacy policy
                </a>
              </p>
            </div>
            <div
              class="about-dialog-right"
            >
              <h2>
                Original project team
              </h2>
              <div
                class="credits-team credits-team-past"
              >
                <div
                  class="team-member"
                >
                  <div
                    class="team-member-mugshot"
                    style="background-image: url(/images/team/dwight.jpg);"
                  />
                  <span
                    class="team-member-name"
                  >
                    Dwight Schrute
                  </span>
                  <span
                    class="team-member-title"
                  >
                    assistant to the regional manager
                  </span>
                </div>
                <div
                  class="team-member"
                >
                  <div
                    class="team-member-mugshot"
                    style="background-image: url(/images/team/michael.jpg);"
                  />
                  <span
                    class="team-member-name"
                  >
                    <a
                      href="https://dunder-mifflin.com"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      Michael Scott
                    </a>
                  </span>
                  <span
                    class="team-member-title"
                  >
                    regional manager
                  </span>
                </div>
              </div>
              <h2>
                Core contributors
              </h2>
              <div
                class="credits-team credits-team-past"
              >
                <div
                  class="team-member"
                >
                  <div
                    class="team-member-mugshot"
                    style="background-image: url(/images/team/jim.jpg);"
                  />
                  <span
                    class="team-member-name"
                  >
                    <a
                      href="http://sabre.com"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      Jim Halpert
                    </a>
                  </span>
                  <span
                    class="team-member-title"
                  >
                    sales
                  </span>
                </div>
                <div
                  class="team-member"
                >
                  <div
                    class="team-member-mugshot"
                    style="background-image: url(/images/team/pam.jpg);"
                  />
                  <span
                    class="team-member-name"
                  >
                    Pam Beesly
                  </span>
                  <span
                    class="team-member-title"
                  >
                    office administrator
                  </span>
                </div>
              </div>
              <div
                class="credits-container"
              >
                <div
                  class="credits-credits-left"
                >
                  <h3>
                    Advisors
                  </h3>
                  <ul>
                    <li>
                      Jan Levinson
                    </li>
                    <li>
                      Robert California
                    </li>
                  </ul>
                  <h3>
                    Additional illustrations
                  </h3>
                  <ul>
                    <li>
                      Kevin Malone
                    </li>
                    <li>
                      Meredith Palmer
                    </li>
                    <li>
                      Stanley Hudson
                    </li>
                  </ul>
                  <h3>
                    Additional code
                  </h3>
                  <ul>
                    <li>
                      Angela Martin
                    </li>
                    <li>
                      Oscar Martinez
                    </li>
                    <li>
                      Phyllis Lapin
                    </li>
                  </ul>
                  <h3>
                    Additional contributors
                  </h3>
                  <ul>
                    <li>
                      Creed Bratton
                    </li>
                    <li>
                      Kelly Kapoor, 
                      <i>
                        customer service
                      </i>
                    </li>
                    <li>
                      Toby Flenderson, 
                      <i>
                        human resources
                      </i>
                    </li>
                  </ul>
                  <h3>
                    Special thanks to
                  </h3>
                  <ul>
                    <li>
                      Andy Bernard, 
                      <i>
                        Cornell
                      </i>
                    </li>
                    <li>
                      Darryl Philbin
                    </li>
                    <li>
                      Kelly Erin Hannon
                    </li>
                    <li>
                      Ryan Howard, 
                      <i>
                        WUPHF
                      </i>
                    </li>
                    <li>
                      <h4>
                        Dunder Mifflin
                      </h4>
                      <ul>
                        <li>
                          David Wallace
                        </li>
                        <li>
                          Nellie Bertram
                        </li>
                        <li>
                          Holly Flax
                        </li>
                      </ul>
                    </li>
                  </ul>
                </div>
                <div
                  class="credits-credits-right"
                >
                  <h3>
                    Translators
                  </h3>
                  <ul>
                    <li>
                      <h4>
                        Arabic
                      </h4>
                      <ul>
                        <li>
                          Charles Miner
                        </li>
                        <li>
                          Karen Filippelli
                        </li>
                      </ul>
                    </li>
                    <li>
                      <h4>
                        German
                      </h4>
                      <ul>
                        <li>
                          Gabe Lewis, 
                          <i>
                            Sabre
                          </i>
                        </li>
                        <li>
                          Jo Bennett, 
                          <i>
                            Sabre
                          </i>
                        </li>
                        <li>
                          Todd Packer
                        </li>
                      </ul>
                    </li>
                  </ul>
                </div>
              </div>
              <div
                class="credits-container"
              >
                <div>
                  <h3>
                    Acknowledgements
                  </h3>
                  <ul>
                    <li>
                      <a
                        href="https://manropefont.com/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Manrope
                      </a>
                       font designed by 
                      <a
                        href="https://gent.media/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Mikhail Shiranda
                      </a>
                      . (
                      <a
                        href="https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=OFL"
                        rel="noopener noreferrer"
                        target="_blank"
                        title="SIL Open Font License, Version 1.1"
                      >
                        License
                      </a>
                      )
                    </li>
                    <li>
                      <a
                        href="https://hubertfischer.com/work/type-rubik"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Rubik
                      </a>
                       font designed by 
                      <a
                        href="https://hubertfischer.com/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Hubert & Fischer
                      </a>
                      . (
                      <a
                        href="https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=OFL"
                        rel="noopener noreferrer"
                        target="_blank"
                        title="SIL Open Font License, Version 1.1"
                      >
                        License
                      </a>
                      )
                    </li>
                    <li>
                      <a
                        href="https://delvefonts.com/fonts/overpass/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Overpass
                      </a>
                       font designed by 
                      <a
                        href="https://delvefonts.com/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Delve Fonts
                      </a>
                      . (
                      <a
                        href="https://github.com/RedHatOfficial/Overpass/blob/master/OFL.txt"
                        rel="noopener noreferrer"
                        target="_blank"
                        title="SIL Open Font License, Version 1.1"
                      >
                        License
                      </a>
                      )
                    </li>
                    <li>
                      Emoji by 
                      <a
                        href="https://openmoji.org/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        OpenMoji
                      </a>
                      . (
                      <a
                        href="https://creativecommons.org/licenses/by-sa/4.0/#"
                        rel="noopener noreferrer"
                        target="_blank"
                        title="Creative Commons Share Alike License 4.0 (CC BY-SA 4.0)"
                      >
                        License
                      </a>
                      )
                    </li>
                    <li>
                      Icons by 
                      <a
                        href="https://fontawesome.com/"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Font Awesome
                      </a>
                      . (
                      <a
                        href="https://fontawesome.com/license/free"
                        rel="noopener noreferrer"
                        target="_blank"
                        title="Creative Commons Share Alike License 4.0 (CC BY-SA 4.0)"
                      >
                        License
                      </a>
                      )
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          class="dialog-primary-action"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
